server/models/annotation.py

```diff
@@ -727,3 +776,109 @@ def validate(self, doc):
         if len(set(elementIds)) != len(elementIds):
             raise ValidationException('Annotation Element IDs are not unique')
         return doc
+
+    def versionList(self, annotationId, user=None, limit=0, offset=0,
+                    sort=[('_version', -1)], force=False):
+        """
+        List annotation history entries for a specific annotationId.  Only
+        annotations that belong to an existing item that the user is allowed to
+        view are included.  If the user is an admin, all annotations will be
+        included.
+
+        :param annotationId: the annotation to get history for.
+        :param user: the Girder user.
+        :param limit: maximum number of history entries to return.
+        :param offset: skip this many entries.
+        :param sort: the sort method used.  Defaults to reverse _id.
+        :param force: if True, don't authenticate the user.
+        :yields: the entries in the list
+        """
+        _versions = set()
+
+        if annotationId and not isinstance(annotationId, ObjectId):
+            annotationId = ObjectId(annotationId)
+        # Don't apply limit and offset here, as they are subject to access
+        # control and other effects
+        entries = self.find(
+            {'$or': [{'_id': annotationId}, {'_annotationId': annotationId}]},
+            sort=sort)
+        isAdmin = force or (user is not None and user['admin'])
+        for entry in entries:
```

Comment:
If I understand this loop correctly, it looks like the reason you aren't using the builtin `filterResultsByPermission` is because of the possibility of duplicate version numbers.  I'm not sure I fully understand the conditions under which this will happen, or why it would be more likely than inconsistencies resulting from overriding methods on `self.collection`.  Would it be possible to avoid the situation alltogether using locks or to use `$group` and `$first` queries to filter out duplicates?

Reply:
Yes.  I'll switch to using an aggregation by group and `filterResultsByPermission`